### ISSUE TITLE
教師あり学習のためのログ案

### ISSUE THREAD

**Issue by og3** [og3/21#12]
## 「正解」の定義
- そのラウンドに勝つための最適な行動
### 状況
- 自分の手札：arry
- 自分の伏せカード：arry
- 自分の手札の合計値：int
- 相手の手札：arry
- 相手の伏せカード：array
- 相手の手札の合計値：int 
- 山札の残り枚数：int
### カードを引いた場合の結果
- カードを引いたらバーストしたか：bool
　こんな感じの教師データを作ったらscikit-learnで読み込んで、次に上記のデータの「状況」のみを渡したら、上記データの「行動」を返すようにしたい
## SPカードはロジックで使わせる
　学習させると複雑になるので。まぁ後々これはSPカードを使うかどうか学習をさせて判断させる。
## 実装手順
- まずは手動でそれっぽいログを10個くらい作ってみて、それが判定できるコードを作る。

**Comment by og3** [comment]
scikit-learnで教師データを読み込み、行動を予測する
データ構造の整理と注意点
ご提示いただいたデータ構造は、scikit-learnで扱う上で非常に適しています。しかし、いくつか注意すべき点と改善点があります。

特徴量（特徴ベクトル）:

数値化: すべての要素を数値化することが重要です。例えば、カードの種類をone-hotエンコーディングなどで数値化します。
正規化: 特徴量間のスケールが異なる場合、標準化や正規化を行うことで、学習の効率が向上します。
重要度の低い特徴量: ターン数や残りの山札など、あまり重要でない特徴量を除外することで、モデルの複雑さを軽減できます。
目的変数:

カードを引いたか: bool値をそのまま使用できます。
scikit-learnを用いた実装例
Python

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

# データを読み込む（例としてCSVファイルの場合）
data = pd.read_csv("game_data.csv")

# 特徴量と目的変数を抽出
X = data[["ターン数", "自分の手札の合計値", ...]]  # 特徴量
y = data["カードを引いたか"]  # 目的変数

# データを訓練データとテストデータに分割
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# モデルを作成し、学習
model = RandomForestClassifier()
model.fit(X_train, y_train)

# 新しいデータで予測
new_data = pd.DataFrame({"ターン数": [3], "自分の手札の合計値": [15], ...})  # 新しいデータ
prediction = model.predict(new_data)
print(prediction)  # 予測結果（カードを引くかどうか）
モデル選択と評価
RandomForestClassifier: 汎用性が高く、多くの場合で良い結果が得られます。
他のモデル:
LogisticRegression: シンプルなモデルで、解釈しやすい
Support Vector Machine (SVM): 高次元データにも対応できる
Neural Network: 大量のデータがある場合に有効
評価指標:
Accuracy: 正確度
Precision: 正の予測の正確さ
Recall: 正のクラスをどれだけ見つけられたか
F1-score: PrecisionとRecallの調和平均
改善点
特徴量エンジニアリング:
相手の行動パターンを特徴量として加える
特定のカードの組み合わせを特徴量として加える
ハイパーパラメータチューニング:
GridSearchCVやRandomizedSearchCVを用いて、最適なパラメータを探索する
アンサンブル学習:
複数のモデルを組み合わせることで、精度向上を目指す
深層学習:
大量のデータがある場合、深層学習を用いてより複雑なモデルを構築できる
その他
不均衡データ: カードを引くか引かないかの割合が偏っている場合、クラスの重み付けを行う
オンライン学習: ゲーム中にリアルタイムで学習し、モデルを更新する
強化学習: 報酬を設定し、試行錯誤を通じて最適な行動を学習させる
まとめ
scikit-learnは、教師あり学習のタスクを簡単に実装できる強力なライブラリです。今回の問題では、RandomForestClassifierを用いて、状況から行動を予測するモデルを構築できます。しかし、より高度なモデルや特徴量エンジニアリングを行うことで、さらに精度を向上させることが可能です。

**Comment by og3** [comment]
## バースト可能性が計算できればいい
　手札の内容はどうでもよくて、プレイヤーは自分と相手のバースト可能性を概算したうえでカードをドローするか判断している。
## バースト可能性の概算
- モンテカルロ法での計算を試す

```python
import random
import numpy as np
from scipy import stats

def monte_carlo_simulation(num_simulations):
    # ... (ゲームのシミュレーションを行う関数)

    results = []
    for _ in range(num_simulations):
        result = play_game()
        results.append(result)

    # バースト確率の区間推定
    confidence_level = 0.95
    mean = np.mean(results)
    std = np.std(results)
    margin_of_error = stats.norm.ppf((1 + confidence_level) / 2) * std / np.sqrt(num_simulations)
    confidence_interval = (mean - margin_of_error, mean + margin_of_error)

    return confidence_interval
```

**Comment by og3** [comment]
# AINPC第一弾
- バースト確率を概算する＝＞そのうえでAIがカードを引くかどうか考えてプレイする、というのがゴール。
# ゲームのルール
## ゲームの盤面について
- 山札（山札となるカードは1～11の数字が書かれたカードが1セット11枚のみ。なので、場に出ているカードと同じカードを引けない。）
- 自分と相手のプレイヤーの手札（1枚を除く、すべてのプレイヤーが見える状態）
- ライフカウンター（初期値は5でマッチに敗北すると減っていく）
## ラウンドの勝敗条件
- 交互に山札から数字が書かれたカードを引き、手持ちカードの合計がより「21」に近い方が勝利となる。
- 手札の合計値が「21」を超えると「バースト」となり、相手も「バースト」していなければ負けとなる。
- お互いの手札の合計値が同じなら引き分け。
## ゲームの勝敗条件
- ライフカウンターの値が0以下になったプレイヤーの負け。
## ゲームの進行
1,ゲーム開始時に山札からカードが2枚ずつ配られる。(1枚目のみ伏せられており、その数字はすべてのプレイヤーが分からない。)
2,プレイヤーのターンとなり、カードを1枚引くかどうかを選択する。
3,相手プレイヤーのターンとなり、同様の選択を行う。
4-1,いずれかのプレイヤーがカードを引く限り、2と3を交互に繰り返す。
4-2,お互いにカードを引かない場合は勝敗判定を行う。
5-1,敗北したプレイヤーのライフカウンターを、そのラウンド数分だけ減らし、お互いにライフカウンターが0でなかったら、次のラウンドとしてこの手順の1からやり直す。
5-2,いずれかのプレイヤーのライフカウンターが0以下になったら、勝利プレイヤーを宣言し、ゲームを終了する。

# やりたいこと
- 自分と相手の手札と残りの山札を見て、バースト確率を概算するロジックを作りたい

# 計算に使うデータの例
```
自分の手札： ["?", 1, 2, 10]
自分の手札の合計値："? + 13"
相手の手札： ["?", 3, 4, 5, 6]
相手の手札の合計値： "? + 18"
山札の残り枚数：2
```

**Comment by og3** [comment]
最初は不確定要素をなくしてやってみる妥協でもいいかも

**Comment by og3** [comment]
1出力で決めたい

**Comment by og3** [comment]
# ゲームのルール
## ゲームの盤面について
- 山札（山札となるカードは1～11の数字が書かれたカードが1セット11枚のみ。なので、場に出ているカードと同じカードを引けない。）
- 自分と相手のプレイヤーの手札（1枚を除く、すべてのプレイヤーが見える状態）
- ライフカウンター（初期値は5でマッチに敗北すると減っていく）
## ラウンドの勝敗条件
- 交互に山札から数字が書かれたカードを引き、手持ちカードの合計がより「21」に近い方が勝利となる。
- 手札の合計値が「21」を超えると「バースト」となり、相手も「バースト」していなければ負けとなる。
- お互いの手札の合計値が同じなら引き分け。
## ゲームの勝敗条件
- ライフカウンターの値が0以下になったプレイヤーの負け。
## ゲームの進行
1,ゲーム開始時に山札からカードが2枚ずつ配られる。(1枚目のみ伏せられており、その数字はすべてのプレイヤーが分からない。)
2,プレイヤーのターンとなり、カードを1枚引くかどうかを選択する。
3,相手プレイヤーのターンとなり、同様の選択を行う。
4-1,いずれかのプレイヤーがカードを引く限り、2と3を交互に繰り返す。
4-2,お互いにカードを引かない場合は勝敗判定を行う。
5-1,敗北したプレイヤーのライフカウンターを、そのラウンド数分だけ減らし、お互いにライフカウンターが0でなかったら、次のラウンドとしてこの手順の1からやり直す。
5-2,いずれかのプレイヤーのライフカウンターが0以下になったら、勝利プレイヤーを宣言し、ゲームを終了する。

# やりたいこと
- 自分と相手の手札と残りの山札を見て、バースト確率を計算したい。
- scikit-learnを使って、バースト確率を出力できるようにしたい。

**Comment by og3** [comment]
## 課題：NPCに個性を持たせるための学習データを作る
　単純に勝つだけならアルゴリズム（ルールベース）で戦うNPCにすればいい。でもそれじゃパターンが読めたらつまらなくなるし、勝てなくなる。
### データのサンプル
　学習に使うデータはこんな感じ。
```json
{
    // abs((21 - my_score) - (21 - opponent_score))
    "score_difference": "int",
    "player_burst_prob": "float",
    "opponent_burst_prob": "float",
    "remaining_cards_num": "int",
    "round": "int",
    "player_life": "int",
    "opponent_life": "int",
    "draw_card": "bool"
}
```
### 学習を通して個性を作る
　データの各値に対して「重みづけ」をするのが個性だと考える。以下はサンプルとして各項目に5段階で重み付けをして個性を分類してみた。1が「気にしない」、5が「重視する」を示す。
　それぞれの重み付けを考慮したアルゴリズムを基に、条件分岐（ルール）を作って学習データを大量に作る。
type | score_difference | player_burst_prob | opponent_burst_prob | remaining_cards_num | round | player_life | opponent_life | memo
-- | -- | -- | -- | -- | -- | -- | -- | --
自己完結型 | 1 | 5 | 3 | 4 | 3 | 5 | 3 | 相手のスコアは気にせず、自分が21に近づくように自分のバースト確率を考慮しつつ、カードを引き続ける。
協調型 | 5 | 3 | 3 | 4 | 1 | 3 | 3 | 相手と似たような動きをし、「接戦」になるように行動を調整する。
自爆待ち型 | 2 | 2 | 5 | 3 | 5 | 2 | 5 | 自分のスコアは気にせず、相手がバーストするまで粘るような戦い方をする。

### データ生成のサンプルコード
```python
import random
import pandas as pd

def generate_game_data(num_samples=10000):
    data = []

    for _ in range(num_samples):
        # ラウンド情報をランダムに生成
        round_number = random.randint(1, 10)
        player_score = random.randint(1, 21)
        opponent_score = random.randint(1, 21)
        player_burst_prob = random.uniform(0, 1)
        opponent_burst_prob = random.uniform(0, 1)
        remaining_cards_num = random.randint(1, 52)  # 山札の残り枚数
        player_life = random.randint(1, 5)
        opponent_life = random.randint(1, 5)

        # スコア差を計算
        score_difference = abs((21 - player_score) - (21 - opponent_score))

        # ルールベースの行動決定
        # 簡易ルール：自分のバースト確率が低く、スコア差が大きい場合は引く
        if player_burst_prob < 0.3 and score_difference > 3:
            draw_card = True
        else:
            draw_card = False

        # データ行を追加
        data.append({
            "score_difference": score_difference,
            "player_burst_prob": round(player_burst_prob, 2),
            "opponent_burst_prob": round(opponent_burst_prob, 2),
            "remaining_cards_num": remaining_cards_num,
            "round": round_number,
            "player_life": player_life,
            "opponent_life": opponent_life,
            "draw_card": draw_card
        })

    return pd.DataFrame(data)

# データ生成
game_data = generate_game_data(10000)

# CSVとして保存する例
game_data.to_csv("game_data.csv", index=False)

# データの表示
print(game_data.head())
```
### データの分割
　訓練データとテストデータに分割。
```python
from sklearn.model_selection import train_test_split

X = game_data.drop("draw_card", axis=1)
y = game_data["draw_card"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```
### モデルの構築
　ここではランダムフォレストを使用。
```python
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)
```
### モデルの評価
　テストデータを使ってモデルを評価。
```python
from sklearn.metrics import accuracy_score

predictions = model.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
print(f"モデルの精度: {accuracy:.2f}")
```
### 予測
　新しいゲーム状況に基づいてAIの行動を予測。これをゲームロジックに組み込んで、カードを引くかどうかの分岐に使う。
```python
new_situation = pd.DataFrame([{
    "score_difference": 5,
    "player_burst_prob": 0.25,
    "opponent_burst_prob": 0.15,
    "remaining_cards_num": 30,
    "round": 2,
    "player_life": 3,
    "opponent_life": 2
}])

prediction = model.predict(new_situation)
print(f"カードを引くか: {'引く' if prediction[0] else '引かない'}")
```

**Comment by og3** [comment]
## 各タイプの具体的な行動
　具体的な行動の条件を分岐で書いて、重み付けとはこれらの個性の発現割合をとして出す。
　例えば、「ディフェンシブ」なパターンを作る場合、ディフェンス7、オフェンス1、協調2としてデータ生成をする。
### ディフェンス
　相手がバーストするのを待つプレイスタイル。
- 相手のバースト確率が30％以下なら引く
- そのラウンドで敗北すると自分のライフが0になる（player_life - round_number == 0）の時、自分のバースト確率が30%以下なら引く
- 山札のカードが5枚以上なら引く
### オフェンス
　相手の手にかかわらず、21に近づくことを目指すプレイスタイル。
- 自分のバースト確率が50％以下なら引く
- スコア差が3以上なら自分のバースト確率が60%以下で引く
- そのラウンドで敗北すると自分のライフが0になる（player_life - round_number == 0）の時、自分のバースト確率が40%以下なら引く
- そのラウンドで勝利すると相手のライフが0になる（opponent_life - round_number == 0）の時、自分のバースト確率が60%以下なら引く
- 山札が3枚以上なら引く
### 協調
　相手のバースト確率に合わせてカードを引くプレイスタイル。
- 自分のバースト確率が50%以下なら引く
- バースト確率の差が20%以下なら引く
- そのラウンドで勝利すると相手のライフが0になる（opponent_life - round_number == 0）の時、自分のバースト確率が40%以下なら引く
- 山札のカードが7枚以上なら引く

**Comment by og3** [comment]
## テストデータの生成例
```
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```
test_size=0.2
→ データの20%をテストデータとして確保し、80%を訓練データにする。

random_state=42
→ 乱数のシードを固定 して、何回実行しても 同じデータ分割結果 になるようにする。

X_train と y_train → モデルを学習するためのデータ
X_test と y_test → モデルの精度を評価するためのデータ
これにより、「見たことのないデータでも正しく予測できるか？」 を確かめることができます。
